### PR TITLE
Display ignored -Xjit and -Xaot args in IGNOREDARGS list

### LINKED ISSUE #19742: Ignored Xjit args do not show up in the IGNOREDARGS list
OpenJ9 will only accept one `-Xjit:<blah-blah>` argument; if more than one `-Xjit:<blah-blah>` argument is presented on the command line, all but one is ignored. However, the ignored `-Xjit` arguments are not included in the `1CIIGNOREDARGS Ignored Args:` list.

It would be nice if ignored `-Xjit` arguments were in the list, so that the user or support team could clearly see that not all the `-Xjit` arguments were actually applied.

### PR DESCRIPTION
**Pull request by luke-li-2003**
The `-Xjit` and `-Xaot` args being ignored because there are multiple `-Xjit` or `-Xaot` arguments (only the last one will be considered), or because there exists `-Xnojit`, `-Xnoaot`, or `-Xint` (ignores the arguments before the in the command line option) will be displayed in the IGNOREDARGS list during the core dump.

Fixes #19742

### CODE REVIEW

File: runtime/rasdump/javadump.cpp
Comment (mpirvu) [inline runtime/rasdump/javadump.cpp:1498]: As far as I know, each defined variable should stay on its own line.
Also see below: "int ignoreXjitBefore = 0, ignoreXaotBefore = 0;" and "int countXjit = 0, countXaot = 0;"

File: runtime/rasdump/javadump.cpp
Comment (mpirvu) [inline runtime/rasdump/javadump.cpp:1517]: To be pedantic I would also check whether `-XX:-MergeCompilerOptions` is specified after `-XX:+MergeCompilerOptions`, invalidating it.

File: runtime/rasdump/javadump.cpp
Comment (mpirvu) [inline runtime/rasdump/javadump.cpp:1542]: I would write that `ignoreLine = !hasXXMerge && (countXjit < numXjit);`

**Review by mpirvu — COMMENTED** [review]
Looks good overall. Please see my inline comments.

**Comment by mpirvu** [comment]
@keithc-ca Could you please review this PR too since it modifies VM files? Thanks

**Review by mpirvu — APPROVED** [review]
LGTM. Waiting for @keithc-ca approval

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1495]: Please
* spell options completely (`-Xjit` instead of `Xjit`)
* the second and subsequent lines of a multi-line comment should start with `*`
* end each sentence with a period
```c
		/* Ignored options for -Xjit and -Xaot follows these rules:
		 * 1. If -Xjit is followed by -Xint or -Xnojit,
		 *    then any -Xjit:optionString beforehand appears as ignored.
		 * 2. If -XX:+MergeCompilerOptions is present,
		 *    then all -Xjit options after -Xint or -Xnojit should not be ignored.
		 * 3. If multiple -Xjit string appear, all but the last one will appear as ignored.
		 *
		 * Similar rules apply for -Xaot.
		 */
```

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1504]: These locals should use `jint` instead of `int` consistent with their uses.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1507]: No space after `*`, please.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1510]: 5 must be 6 here - we don't want to match other options that start with `-Xint`. I think the same issue applies to `-Xnojit` `-Xnoaot`, `-XX:+MergeCompilerOptions` and `-XX:-MergeCompilerOptions`.

Please finish each sentence with a period.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1523]: Please spell the option names in full in comments.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1527]: Please don't use `++` at the statement level:
```c
				numXjit += 1;
```

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1548]: Testing `countXjit < numXjit` is redundant here in the `else` block: please just remove it:
```c
					ignoreLine = !hasXXMerge;
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]
I think the code should be reorganized so ignored options are printed in the same order they're specified on the command line. I expect that means the first new loop must be inserted before the existing loop, and the work of the second new loop must be merged into the existing loop.

File: runtime/rasdump/javadump.cpp
Comment (mpirvu) [reply runtime/rasdump/javadump.cpp:1548]: I think we still need the `(countXjit < numXjit)` test.
E.g. "-Xjit:opt1 -Xint -Xjit:opt2 -Xjit:opt3"
The test `if (countXjit <= ignoreXjitBefore)` will result in ignoring of -Xjit:opt1.
The test `(countXjit < numXjit)`  will result in ignoring of -Xjit:opt2, but not -Xjit:opt3

Comment (keithc-ca) [reply runtime/rasdump/javadump.cpp:1548]: Sorry, yes, that's right - I misread the conditions.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1536]: The type of these variables should be `jint`, not `int`. They should be incremented with `+= 1`.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1538]: I suggest a better name is `ignored` (for consistency with `anyIgnored`).

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1491]: Missing `-` before `XX`.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1510]: ignores -> ignore (see similar comments below).

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1524]: The final `*/` of a multi-line comment should be on a line by itself.
```c
				/* If -XX:-MergeCompilerOptions is encountered after
				 * -XX:+MergeCompilerOptions, the former is overriden.
				 */
```

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1520]: Missing period; also lines 1527, 1530.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1548]: Case -> Cases; `*/` on the next line:
```c
					/* Cases 2 & 3: Only the last option is not ignored
					 * unless -XX:+MergeCompilerOptions.
					 */
```
Similar comment for lines 1557-1558.

**Review by keithc-ca — CHANGES_REQUESTED** [review]
My comment about printing ignored options in command-line order remains.

File: runtime/rasdump/javadump.cpp
Comment (luke-li-2003) [reply runtime/rasdump/javadump.cpp:1520]: I want to note that this does break the existing formatting style of the file, which does not use period (e.g. line 1471). I hope this will not be an issue.

Comment (keithc-ca) [reply runtime/rasdump/javadump.cpp:1520]: I consider line 1471 to contain a "sentence fragment" (it neither starts with a capital nor ends with a period). This comment starts with a capital so it should have the matching period. This comment could be written as a fragment as well. So either
```c
				/* Case 2. */
```
or
```c
				/* case 2 */
```
but not
```c
				/* Case 2 */
```
or
```c
				/* case 2. */
```

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1492]: Rather than counting `-Xjit` and `-Xaot` options, we could just keep track of the index of the last of each such option which should simplify the code to decide which options are ignored.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by luke-li-2003** [comment]
All the comments should have been addressed now

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1486]: I don't see how "in args->nOptions" makes sense here (it's an integer, not something that could contain options).

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1516]: Missing space before `{`, here and on lines 1521, 1534, and 1539.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1536]: Suggest:
* using `if` so the variable is not reassigned `false` (which isn't helpful)
* reversing the right-hand-side of `||` to make a set of parentheses clearly redundant
```c
				if ((i < lastXnojitOrXint) || ((i < lastXjit) && !hasXXMerge)) {
					optionIgnored = true;
				}
```
Same suggestion for lines 1540-1541.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1544]: This should be tested independently of the code above (in line 1547 as I suggested earlier): an `-Xjit` or `-Xaot` option might be not consumed but not considered ignored above.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by luke-li-2003** [comment]
All the comments should have been addressed now

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [reply runtime/rasdump/javadump.cpp]: Please improve this comment. I suggest simply removing "in args->options" unless I misunderstand its intent.

File: runtime/rasdump/javadump.cpp
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:1547]: This comment is wrong: the argument is not consumed. I repeat my suggestion of putting this test in line 1551 (which only needs to occur if we haven't already decided the option is ignored).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
Please address the failure of the line endings check.

**Comment by luke-li-2003** [comment]
Trailing whitespace removed

**Comment by keithc-ca** [comment]
Jenkins test sanity amac jdk17

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**
